### ISSUE TITLE
Bug about the weight factor alpha of focal loss 

### ISSUE THREAD

**Issue by shuaihuaiyi** [kornia/kornia#875]
## 🐛 Bug

https://github.com/kornia/kornia/blob/cf8e854546110e3dac1caa35449429fc7af328e5/kornia/losses/focal.py#L86

Hi, I find someting wrong about the multi-class focal loss computation.

It seems like we are multiplying the loss of different classes by a same alpha factor. I think this is equal to just reduce lr, and cannot address the class imbalance issue. In the original paper, alpha is defined as another cased function just like class probability (Eq.2). 

Please take a look about it, thank you !

**Comment by Zhylkaaa** [comment]
Up! I also think that this implementation is not correct original implementation in fvcore uses binary cross entropy and multiplies class 0 by alpha and class 1 by 1-alpha, so alpha is equivalent to weight parameter in CrossEntropy and I believe should also be normalized. (btw. in binary cross entropy you actually have this implemented right)

**Comment by shijianjian** [comment]
Yes. Alpha is a term that used as the weight in crossentropy and our implementation is probably wrong. @Zhylkaaa Are you willing to assist benchmarking against standard libraries like [this](https://github.com/qubvel/segmentation_models.pytorch/blob/a542c55575980aa19076bc53595c86df2b4103e3/segmentation_models_pytorch/losses/_functional.py#L36) then to throw a PR for a fix?

**Comment by Zhylkaaa** [comment]
hey @shijianjian, my implementation is more about multi-class situation, so I don't know if there is a direct comparison. But I can try to help (but I am afraid it can take some time, it is a very busy month 😁).


Generally my implementation should be equivalent to qubvel, but you should expend single dimensional output to 2 dimensional vector (just [output, 1-output] for each prediction) , and set alpha to 1-alpha (in my implementation alpha is a weight for class 0)

**Comment by shijianjian** [comment]
> hey @shijianjian, my implementation is more about multi-class situation, so I don't know if there is a direct comparison. But I can try to help (but I am afraid it can take some time, it is a very busy month 😁).

Great. Thanks for your help.

Personally, I set the alpha value as per the class ratio of examples in a dataset. Basically, lower the number of examples, higher the weight as in ```nn.CrossEntropy(weight)```. Thus, I believe it might make sense to expose the ```alpha``` as both a tensor and a float. In such case, a float alpha would mean to handle binary classification while a tensor can be for both binary or multi-class. What do you think?

Another option can be the implementation of [torch-toolbelt](https://github.com/BloodAxe/pytorch-toolbelt/blob/920e03876805351ed5645e439a64074cb4f37589/pytorch_toolbelt/losses/functional.py#L82) which totally ignored the ```alpha``` for the multiclass cases.

**Comment by Zhylkaaa** [comment]
Actually my implementation is exactly what you said. If you pass float to alpha it will be used as what I called one-vs-other mode (alpha is the weight for class 0, while 1-alpha is weight for the rest of classes) and if you pass a list (Iterable) than it will be used as normal weights (like in `nn.CrossEntropyLoss`)
This is implemented hear https://github.com/Zhylkaaa/simpletransformers/blob/add_losses/simpletransformers/losses/focal_loss.py#L100

**Comment by VladShtompel** [comment]
@edgarriba I'd like to have a go at it, implementing a fix like what shijianjian and Zhylkaaa suggested.

What do you think?

**Comment by edgarriba** [comment]
Sure , please go ahead

**Comment by ducha-aiki** [comment]
Is it still the case?
